### PR TITLE
 ENH: Raise ValueError for unsupported Window functions

### PR DESCRIPTION
**Pull request by ihsansecer**
- [x] ref #26597
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Weighted std could be implemented but there are also other functions weighted `Window` is lacking.

### CODE REVIEW

**Review by jreback — CHANGES_REQUESTED** [review]
this seems like overkill 

when you create a Window it already has restricted methods

what error message does this have now (before this PRA)?

**Comment by ihsansecer** [comment]
@jreback it is ok when directly calling an unsupported function but raised error (`AttributeError: 'Window' object has no attribute 'dtype'`) is not informative for `agg` function

File: pandas/core/window.py [hunk 1]
```diff
@@ -659,6 +660,26 @@ def validate(self):
         else:
             raise ValueError("Invalid window {0}".format(window))
 
+    def _validate_aggregate(self, arg, *args, **kwargs):
+        # supported weighted window functions
```
Comment (jreback) [inline pandas/core/window.py]: rather that this (listing methods)

you can look a bit more into where the methods are called and catch an invokation error (AttributError) for example

we already have quite a bit of code to facilitate this and don’t want to add more layers

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/base.py [hunk 2]
```diff
@@ -603,9 +609,6 @@ def _aggregate_multiple_funcs(self, arg, _level, _axis):
                     keys.append(col)
                 except (TypeError, DataError):
                     pass
-                except ValueError:
```
Comment (ihsansecer) [inline pandas/core/base.py]: I am not sure what this is for just checked some tests locally and didn't fail

Comment (WillAyd) [reply pandas/core/base.py]: Hmm a little wary here - this does show as hit on code coverage:

https://codecov.io/gh/pandas-dev/pandas/src/master/pandas/core/base.py#L608

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/core/base.py (hunk 2 above)
Comment (ihsansecer) [reply pandas/core/base.py]: It seems this is for nuisance columns

File: pandas/core/base.py [hunk 3]
```diff
@@ -314,7 +314,13 @@ def _try_aggregate_string_function(self, arg, *args, **kwargs):
...
-            return f(self, *args, **kwargs)
+            try:
+                return f(self, *args, **kwargs)
+            except AttributeError:
+                raise ValueError(
+                    "'{arg}' is not a valid function for "
+                    "'{cls}' object".format(arg=arg, cls=type(self).__name__)
+                )
 
         raise ValueError("{arg} is an unknown string function".format(arg=arg))
```
Comment (WillAyd) [inline pandas/core/base.py]: Does this get hit any more?

Comment (ihsansecer) [reply pandas/core/base.py]: It does for single string argument or a dictionary. Other one is for list or tuple with invalid functions

File: pandas/core/base.py [hunk 4]
```diff
@@ -554,7 +560,19 @@ def is_any_frame():
...
         elif is_list_like(arg):
             # we require a list, but not an 'str'
-            return self._aggregate_multiple_funcs(arg, _level=_level, _axis=_axis), None
+            try:
+                return (
+                    self._aggregate_multiple_funcs(arg, _level=_level, _axis=_axis),
+                    None,
+                )
+
+            # if no results
```
Comment (jreback) [inline pandas/core/base.py]: can you give a more comprehensive comment here.

Why are we raising ValueError (and not AttributeError from the inner function)?

**Review by jreback — CHANGES_REQUESTED** [review]
looks good, question about if we are correctly raising a ValueError in the list-like case (which you are catching).

File: pandas/core/base.py (hunk 4 above)
Comment (ihsansecer) [reply pandas/core/base.py]: I guess AttributeError would be a better option if not inconvenient since ValueErrors are conflicting. Will change it

**Comment by ihsansecer** [comment]
Please don't merge this as it raises `ValueError: no results` in some cases. Will check it when I have time

**Review by WillAyd — APPROVED** [review]
lgtm ex any @jreback comments and any CI failures

**Comment by ihsansecer** [comment]
`skew` and `kurt` functions as an argument to `agg` raises `unknown string function` error. `min` and `max` functions on the other hand doesn't raise anything and returns `Window`object.

File: pandas/core/base.py [hunk 5]
```diff
@@ -314,9 +314,16 @@ def _try_aggregate_string_function(self, arg, *args, **kwargs):
 
         f = getattr(np, arg, None)
         if f is not None:
-            return f(self, *args, **kwargs)
+            try:
+                return f(self, *args, **kwargs)
+
+            except (AttributeError, TypeError):
+                raise AttributeError(
```
Comment (jreback) [inline pandas/core/base.py]: I would pass here, then just use a single raise (line 326)

File: pandas/core/window.py [hunk 6]
```diff
@@ -511,6 +511,7 @@ class Window(_Window):
         Set the labels at the center of the window.
     win_type : str, default None
         Provide a window type. If ``None``, all points are evenly weighted.
+        Other types are only applicable for `mean` and `sum` functions.
```
Comment (jreback) [inline pandas/core/window.py]: this *should* be on .agg so I wouldn't add anything here

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/window.py (hunk 6 above)
Comment (ihsansecer) [reply pandas/core/window.py]: `agg` is using `shared_docs`. I was avoiding to copy it and make a single line change. Will remove it only if no objection

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm @ihsansecer ping on green.

**Comment by jreback** [comment]
thanks @ihsansecer

**Comment by jreback** [comment]
btw now that pandas/tests/window/ exists would love to split test_window even more :->

**Comment by ihsansecer** [comment]
@jreback noted :)

**Pull request merged**